Avahi hostname rename to <host>-2.local after boot — WiFi+eth0 dual-publish window cached by LAN mDNS

## Symptom

`device-smoke.sh --both` (or `--server`) on a snapMULTI server reports:

```
[ERROR] Snapcast mDNS not visible for <hostname>.local — check snapserver container + avahi-daemon
[WARN] AirPlay mDNS not visible for <hostname>.local — service may be disabled or no client has paired yet
```

Yet external clients connect and stream audio fine. fleet-smoke from another host (e.g. macOS operator) auto-discovers the server via `_snapcast._tcp` without trouble. The smoke check is reporting a real avahi state, but the user-visible impact is zero because the SRV targets still resolve to the right IP — just under the wrong hostname.

## Live evidence (snapvideo, v0.7.6, 2026-05-16)

`avahi-browse -prt _snapcast._tcp` on the host:

```
=;eth0;IPv4;Snapcast;_snapcast._tcp;local;snapvideo-2.local;192.168.63.4;1704;
```

Note the `snapvideo-2.local` SRV target instead of `snapvideo.local`. `getent hosts snapvideo.local` still resolves to `192.168.63.4` via local NSS, but the mDNS-advertised hostname for `_snapcast._tcp` and `_raop._tcp` is the renamed one. AirPlay (`_raop._tcp`) and Spotify (`_spotify-connect._tcp`) inherit the same SRV target.

## Smoking gun in journal

```
15:51:26  avahi-daemon: Registering new address record for fe80::dea6:32ff:feb4:c617 on wlan0.*.
15:51:26  avahi-daemon: Registering new address record for 192.168.63.81 on wlan0.IPv4.
15:51:26  avahi-daemon: Registering new address record for fe80::dea6:32ff:feb4:c616 on eth0.*.
15:51:26  avahi-daemon: Registering new address record for 192.168.63.4 on eth0.IPv4.
15:51:27  avahi-daemon: Server startup complete. Host name is snapvideo.local. Local service cookie is 1148815785.
15:51:32  avahi-daemon: Withdrawing address record for fe80::dea6:32ff:feb4:c617 on wlan0.
15:51:32  avahi-daemon: Withdrawing address record for 192.168.63.81 on wlan0.
                                                                            
...32 min later...
                                                                            
16:23:56  systemd[1]: Starting systemd-timedated.service - Time & Date Service...
16:23:56  avahi-daemon: Withdrawing workstation service for vethfbf2005/br-…/docker0/wlan0/eth0/lo.
16:23:56  avahi-daemon: Withdrawing address record for fe80::dea6:32ff:feb4:c616 on eth0.
16:23:56  avahi-daemon: Host name conflict, retrying with snapvideo-2
16:23:57  avahi-daemon: Server startup complete. Host name is snapvideo-2.local.
```

## Root cause

A 6-second window at boot (`15:51:26 → 15:51:32`) where avahi-daemon advertises the canonical hostname on BOTH wlan0 (192.168.63.81) and eth0 (192.168.63.4) before `boot-tune.sh` disables WiFi. The LAN mDNS cache (in our environment: the fritz.box router) memorizes the dual publication.

32 minutes later, `systemd-timedated.service` starts (likely a dbus call after a clock event) and avahi performs a withdraw / re-probe cycle. On re-probe avahi queries the network for `<hostname>.local` conflict — the router still has the dual publication cached → conflict → rename to `<hostname>-2`.

Same avahi PID throughout the rename: not a restart, just an in-process withdraw/re-register cycle.

The v0.7.5 `tune_nm_docker_unmanaged()` fix prevents NM from adopting Docker bridges as managed interfaces — orthogonal to this scenario. The PartOf= removal in v0.7.5 prevents avahi restart cascades to snapmulti-server — also orthogonal.

## User-visible impact

**None today on snapvideo**: 5/6 clients connected, audio playing, fleet-smoke from macOS finds the server. The smoke false-positive is the only operator-visible artefact.

But the rename is real avahi state that breaks:
- canonical hostname for AirPlay device name (clients see \"snapvideo-2 AirPlay\" instead of \"snapvideo AirPlay\")
- Spotify Connect zeroconf device name (`hostname` derived)
- any operator tool that browses mDNS and matches on the canonical hostname

Sufficient for community-launch ship, but a latent fragility we want to close in v0.7.7.

## Reproduction

1. Reflash a server-mode device whose Pi has working WiFi auto-association (Pi 4 / Pi 5 with credentialled wpa_supplicant).
2. Connect Ethernet too.
3. Observe avahi-daemon startup logs — both wlan0 and eth0 register before `boot-tune.sh` runs `nmcli device disconnect wlan0`.
4. Force the LAN cache to retain by `avahi-browse -t _services._dns-sd._udp` from another host during the 6-second window.
5. Wait for any process to dbus-trigger `systemd-timedated.service` (NTP clock skew correction is a common one).
6. Observe rename in journal.

Local mitigation today: `sudo systemctl restart avahi-daemon snapmulti-server.service && docker restart shairport-sync librespot tidal-connect mpd` reclaims the canonical hostname after the router cache flushes.

## Mitigation options for v0.7.7

| Option | Approach | Trade-off |
|--------|----------|-----------|
| **A — Order dependency** | `Override Unit avahi-daemon.service` with `After=snapmulti-boot-tune.service` (and `Requires=` to fail-stop on tune failure). boot-tune kills WiFi before avahi ever publishes. | Cleanest. Adds ~3-5 s to mDNS-publish wall time. Risk: if boot-tune fails for any reason, avahi never publishes. Should pair with `Requisite=` not `Requires=` so we don't trap into a wedge. |
| **B — Avahi interface whitelist** | `tune_avahi_daemon` writes `allow-interfaces=eth0` when Ethernet is the chosen path; `interfaces-deny=` for Docker bridges. | Surgical. avahi never advertises on wlan0 at all when eth0 is up. Doesn't depend on ordering. |
| **C — Post-boot-tune flush** | `boot-tune.sh` does `systemctl reload-or-restart avahi-daemon` after WiFi withdraw to force a clean republish before the LAN cache settles. | Doesn't prevent the 6-sec window, just narrows it. Race against router cache TTL. |
| **D — Document + smoke softening** | Mark as known issue; make `check_mdns.sh` accept `<hostname>(-N)?\\.local` as valid SRV target. | Cosmetic only. Doesn't address the underlying rename. |

Recommendation: **B** as the structural fix (smallest blast radius, no boot ordering changes, easy to test in CI by mocking `iface_chosen=eth0`), with optional **D** as a smoke-check polish in the same PR.

## Acceptance

- Repeat reproduction → no rename after 32 min or after manual `systemctl start systemd-timedated`
- `avahi-browse -prt _snapcast._tcp` SRV target is `<hostname>.local` (not `<hostname>-2.local`) one hour after boot
- `device-smoke.sh --server` clean on snapvideo + snapdigi + snap-fleet without intervention
- New test: `tests/test_avahi_interface_scope.sh` asserting `allow-interfaces=eth0` is written when Ethernet is up at config-time
- Smoke check `check_mdns.sh` keeps strict mode by default but accepts `<hostname>-N.local` with an explicit info note (so rename detection is still surfaced for diagnostics)

## Out of scope

- Multi-server snapMULTI fleets where two devices share a hostname (different problem)
- avahi rename in containerised mDNS publish (libavahi-client) — that path uses the host's avahi-daemon, so a host-side fix covers it
- WiFi-only installs (no Ethernet) — those don't have the dual-publish window

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+- **Avahi wired-priority defeats the `<host>-2.local` rename race (#425)** — on installs where the user configured WiFi credentials *and* Ethernet is plugged at first boot, `tune_avahi_daemon` previously enumerated both interfaces from the default route table during the 6-second window before `boot-tune.sh` disabled WiFi. Both wlan0 and eth0 momentarily published the same `<host>.local`, and LAN mDNS reflectors (e.g. FritzBox) cached the WiFi address; when eth0 then claimed the same name, Avahi negotiated a conflict and renamed to `<host>-2.local` (observed live on snapvideo 2026-05-16). `tune_avahi_daemon` now checks `/sys/class/net/eth*` and `/sys/class/net/en*` for `carrier=1` first: if a wired interface has carrier we restrict `allow-interfaces=<eth>` to wired only, so the reflector never sees a transient WiFi announcement to cache. WiFi-only devices (no wired carrier) fall through to the existing default-route + IPv4-up enumeration, unchanged. Defensive — does not change the steady-state behaviour after `boot-tune.sh` disables WiFi, but eliminates the first-boot race window. `tests/test_avahi_readiness.sh` gains 3 assertions covering the `/sys/class/net` enumeration, carrier check, and `wired_iface` precedence.
+
 ## [0.7.6] — 2026-05-16
 
 > Consolidates changes from v0.7.4 (2026-05-13) and v0.7.5 (2026-05-15) which shipped without separate CHANGELOG sections — see per-tag [GitHub Releases](https://github.com/lollonet/snapMULTI/releases) for the tag-level narrative.

--- a/scripts/common/system-tune.sh
+++ b/scripts/common/system-tune.sh
@@ -454,26 +454,46 @@ tune_avahi_daemon() {
     # Restrict to the primary physical interface. On Ethernet+WiFi
     # devices, Avahi can otherwise publish the same hostname on the
     # transient WiFi DHCP address before WiFi exclusivity disables wlan0;
-    # macOS then keeps the stale .local address in cache.
+    # macOS and consumer mDNS reflectors (e.g. FritzBox) then keep the
+    # stale .local address in cache, which forces a hostname rename to
+    # <host>-2.local on conflict (issue #425).
+    #
+    # Wired-carrier priority: if any eth*/en* interface has carrier we
+    # restrict Avahi to wired only — this mirrors boot-tune.sh which
+    # disables WiFi when Ethernet has an IP, but takes effect during the
+    # first-boot window before boot-tune has run. WiFi-only devices (no
+    # eth carrier) fall through to the default-route enumeration.
     local ifaces=""
     local iface
-    for iface in $(ip -o route show default 2>/dev/null | awk '
-        {
-            for (i = 1; i <= NF; i++) {
-                if ($i == "dev" && $(i + 1) ~ /^(eth|wlan|en)/ && !seen[$(i + 1)]++) {
-                    print $(i + 1)
+    local wired_iface=""
+    for iface in /sys/class/net/eth* /sys/class/net/en*; do
+        [[ -e "$iface" ]] || continue
+        if [[ "$(cat "$iface/carrier" 2>/dev/null)" == "1" ]]; then
+            wired_iface="${iface##*/}"
+            break
+        fi
+    done
+    if [[ -n "$wired_iface" ]]; then
+        ifaces="$wired_iface"
+    else
+        for iface in $(ip -o route show default 2>/dev/null | awk '
+            {
+                for (i = 1; i <= NF; i++) {
+                    if ($i == "dev" && $(i + 1) ~ /^(eth|wlan|en)/ && !seen[$(i + 1)]++) {
+                        print $(i + 1)
+                    }
                 }
             }
-        }
-    '); do
-        [[ -n "$ifaces" ]] && ifaces="${ifaces},"
-        ifaces="${ifaces}${iface}"
-    done
-    if [[ -z "$ifaces" ]]; then
-        for iface in $(ip -o -4 addr show scope global up 2>/dev/null | awk '$2 ~ /^(eth|wlan|en)/ && !seen[$2]++ {print $2}'); do
+        '); do
             [[ -n "$ifaces" ]] && ifaces="${ifaces},"
             ifaces="${ifaces}${iface}"
         done
+        if [[ -z "$ifaces" ]]; then
+            for iface in $(ip -o -4 addr show scope global up 2>/dev/null | awk '$2 ~ /^(eth|wlan|en)/ && !seen[$2]++ {print $2}'); do
+                [[ -n "$ifaces" ]] && ifaces="${ifaces},"
+                ifaces="${ifaces}${iface}"
+            done
+        fi
     fi
     if [[ -n "$ifaces" ]]; then
         if ! grep -q "^allow-interfaces=${ifaces}$" "$conf"; then

--- a/tests/test_avahi_readiness.sh
+++ b/tests/test_avahi_readiness.sh
@@ -232,6 +232,20 @@ assert 'grep -qF "ip -o route show default" "$SYSTEM_TUNE_SH"' \
 assert 'grep -qF "ip -o -4 addr show scope global up" "$SYSTEM_TUNE_SH"' \
        'tune_avahi_daemon has IPv4-up physical-interface fallback'
 
+# #425 — wired-carrier priority prevents the dual-publish race during
+# the 6-sec window where wlan0 still has its DHCP address before
+# boot-tune.sh disables WiFi. When eth*/en* has carrier we restrict
+# Avahi to wired only, so the reflector never sees a transient WiFi
+# announcement that would force a <host>-2.local rename on conflict.
+assert 'grep -qF "for iface in /sys/class/net/eth* /sys/class/net/en*" "$SYSTEM_TUNE_SH"' \
+       'tune_avahi_daemon enumerates wired interfaces by /sys/class/net'
+
+assert 'grep -qF "carrier" "$SYSTEM_TUNE_SH"' \
+       'tune_avahi_daemon checks wired carrier state'
+
+assert 'grep -qF "wired_iface" "$SYSTEM_TUNE_SH"' \
+       'tune_avahi_daemon prefers wired interface when carrier is up'
+
 echo
 echo "=== firstboot.sh — Avahi retune after network ==="
 

